### PR TITLE
add comment feature to webiste

### PR DESCRIPTION
**Pull request by chuckluv**

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/data/Post.java
Comment (KoopaKing) [inline portfolio/src/main/java/com/google/sps/data/Post.java]: Did you run this through the formatter? You're missing a space before the bracket.

File: portfolio/src/main/java/com/google/sps/data/Post.java
Comment (KoopaKing) [inline portfolio/src/main/java/com/google/sps/data/Post.java:28]: newline for EOF

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java
Comment (KoopaKing) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Trailing whitespace? gross.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java
Comment (KoopaKing) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Might be nice to do some validation here.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java
Comment (KoopaKing) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: remove this.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java
Comment (KoopaKing) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: should this be a different path, like say /comments or something?

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java
Comment (KoopaKing) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: You're using a bunch of magic strings. Please refactor them.

Also by this I mean your code depends on the spelling of "Comments", "timestamp", and "comment" all matching.

it would be better to have something like:
```java
  private static final String COMMENTS_ENTITY_NAME = "Comments";
  private static final String TIMESTAMP_PROPERTY_KEY = "timestamp";
  private static final String COMMENT_PROPERTY_KEY = "comment";
```

then use those constants in both places.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java
Comment (KoopaKing) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Did they tell you to put the redirect in the rest API response? That's pretty strange for a successful response.

File: portfolio/src/main/webapp/script.js
Comment (KoopaKing) [inline portfolio/src/main/webapp/script.js]: remove this before merging please.

File: portfolio/src/main/webapp/script.js
Comment (KoopaKing) [inline portfolio/src/main/webapp/script.js]: how about `comments` instead of `com`?

**Review by KoopaKing — CHANGES_REQUESTED** [review]
/formatter-meme

**Comment by chuckluv** [comment]
http://charleslovett-step-2020.appspot.com/
This is the url for the webpage so you can look at it

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java
Comment (KoopaKing) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: This is not the `COMMENT_PROPERTY_KEY`... It happens to have the same value, but it's not the same. One is a storage layer concept, one is a user-facing API concept. This is something more like `COMMENT_REQUEST_PARAMETER_KEY`.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java
Comment (KoopaKing) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: I think you can put the limit here... Something like:
```java
FetchOptions fetchOptions = FetchOptions.Builder.withLimit(max);
for (Entity entity : results.asIterable(fetchOptions)) {
   ...
```

Then you wont need to do this whole count check thing...

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java
Comment (KoopaKing) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: What is this `maxnum` thing? Or just as importantly, why does it exist?

Also, without that, some of the below code doesn't make sense.

**Review by KoopaKing — APPROVED** [review]
Seems ok to merge, but try to address the comments first.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java
Comment (bsidhom) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:26]: Please explicitly import just the classes you use. This makes it easier to tell at a glance where different classes come from and what your dependencies are.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java
Comment (bsidhom) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:35]: A more descriptive name will aid readability here, e.g., `maxResultsPerQuery` (or whatever makes sense to you).

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java
Comment (bsidhom) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:46]: Javadoc comments say that this specifically does not use a transaction. Is that safe here? What happens if somebody else adds a comment as we're listing them here? (Note: I don't know the answer to this, just asking whether you've investigated.)

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java
Comment (bsidhom) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:53]: Is it worth checking that this property exists and checking that it's the correct type before trying to cast?

I'm not sure how AppEngine translates inline exceptions (e.g., NullPointerExceptions) to responses (500s?).

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java
Comment (bsidhom) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:78]: Can a user specify _both_ a comment and a max number of comments in the same request?

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java
Comment (bsidhom) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:83]: Note: this generates a local system (AppEngine) log and swallows the exception. Do you want to surface this error to the user in any way?

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java
Comment (bsidhom) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:60]: According to the [javadoc](https://javadoc.io/doc/com.google.code.gson/gson/latest/com.google.gson/com/google/gson/Gson.html), this class is thread-safe. You could store a Gson instance in the constructor and reuse it across requests. This is especially helpful if you end up adding configuration options and using Gson in multiple places.

File: portfolio/src/main/webapp/script.js
Comment (bsidhom) [inline portfolio/src/main/webapp/script.js:42]: This could use a more descriptive name (`createCommentElement`?).

File: portfolio/src/main/webapp/script.js
Comment (bsidhom) [inline portfolio/src/main/webapp/script.js:70]: Looks like this is only used for comments right now. If there are plans to extend this and make it work with arbitrary "tasks", this name might make sense. Otherwise, consider renaming to `deleteComment` or similar.

File: portfolio/src/main/java/com/google/sps/servlets/DeleteServlet.java
Comment (bsidhom) [inline portfolio/src/main/java/com/google/sps/servlets/DeleteServlet.java:39]: It looks like this [throws an exception](https://cloud.google.com/appengine/docs/standard/java/javadoc/com/google/appengine/api/datastore/DatastoreService.html#delete-com.google.appengine.api.datastore.Key...-) if the given key is invalid. I assume that non-existent (e.g., already-deleted) keys are considered invalid. In that case, it's possible for legitimate requests from different users to result in `IllegalArgumentException`s.

As I mentioned above, I'm not sure how AppEngine handles uncaught exceptions, but you may want to consider at least catching exceptions and trying to return the appropriate response codes in that case (unless we'll be addressing that later).

**Review by bsidhom — APPROVED** [review]
Not blocking merge on any of these comments (some are nits), but I think we should at least talk about error handling and what we expect to happen in edge cases.

**Pull request merged**
